internal/controllers/scheduling/op.go

```diff
@@ -0,0 +1,233 @@
+package scheduling
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"hash/fnv"
+	"reflect"
+	"time"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type op struct {
+	Synthesizer *apiv1.Synthesizer
+	Composition *apiv1.Composition
+	Reason      opReason
+
+	Dispatched time.Time
+
+	id               uuid.UUID // set when patch is built
+	synthRolloutHash []byte    // memoized
+}
+
+func newOp(synth *apiv1.Synthesizer, comp *apiv1.Composition) *op {
+	syn := comp.Status.CurrentSynthesis
+	o := &op{Synthesizer: synth, Composition: comp}
+
+	switch {
+	case comp.DeletionTimestamp != nil || !comp.InputsExist(synth) || comp.InputsOutOfLockstep(synth) || !controllerutil.ContainsFinalizer(comp, "eno.azure.io/cleanup"):
```

Comment:
A few questions on this one:
- Is the finalizer guard here to protect against compositions that haven't been seen by Eno yet or ones whose finalizer has been just removed? If it's the former (which is my guess) what goes wrong when we don't guard it?
- Is our reliance on the input controller and the fact that this could get stuck the reason for the op timeout in the controller?

Reply:
> Is the finalizer guard here to protect against compositions that haven't been seen by Eno yet or ones whose finalizer has been just removed? If it's the former (which is my guess) what goes wrong when we don't guard it?

I think this would technically be safe without guarding on the finalizer, since swapping the state doesn't necessarily mean there will be anything to clean up on deletion. The synthesis lifecycle controller will add the finalizer before creating the pod anyway.

For whatever reason removing this check caused one of the batch reconciliation tests to fail so here it stays 🤷 .